proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+      module ``N`` within ``M`` with export QEnv ``N_E``, the
+      export QEnv of ``M`` includes ``qualify(N, N_E)``.
+
+   #. A local module definition for ``M`` adds entries to the ambient QEnv
+      ``E0`` of the enclosing module.
+      Let the export QEnv of ``M`` be ``X``; then ``E0`` is extended
+      with the environment ``qualify(M, X)``. If the ``qualified`` keyword is missing from
+      the definition of the local module, then ``X`` is additionally added to ``E0``.
+
+      Example::
+
+        module Outer where
+          module Inner where
+            five = 5
+
+      The export QEnv of ``Inner`` is ``{five ↦ (Outer, Inner.five)}``.
+      The ambient QEnv of ``Outer`` is thus extended with ``{five ↦ (Outer, Inner.five),
+      Inner.five ↦ (Outer, Inner.five), Outer.Inner.five ↦ (Outer, Inner.five)}``. The
+      first of these comes from adding the export QEnv (call it ``X``, as above)
+      of ``Inner`` to the ambient QEnv
+      in ``Outer``, because the declaration for ``Inner`` does not say ``qualified``.
+      The second of these comes from adding ``qualify(Inner, X)``. The third comes
+      from the rule in the Haskell 2010 Report (Section 5.5.1) that new names are
+      brought into scope both qualified and unqualified.
+
+#. A new declaration form is introduced::
+
+     decl ::= ... | 'import' 'module' modids [ impspec ]
+
+   The declaration is allowed in ``let`` and ``where`` clauses. It
+   modifies the environment for the scope of the ``let`` / ``where``
+   block. For instance in
+
+   ::
+
+     let
+       x = u
+       import module A
+     in
+     v
+
+   The environment is modified for ``u`` and ``v``.
+
+   Let ``E`` be the ambient QEnv. Let ``I`` be the interpretation
+   of ``impspec`` with respect to the environment ``strip(modids, E)``. (If ``impspec``
+   is missing, then let ``I = strip(modids, E)``.)
+   The declaration ``import module modids (impspec)`` modifies the ambient QEnv
+   to become ``E ∪ I``.
+
+   Note that the declaration form includes the word ``module`` to distinguish
+   it from a normal ``import`` which induces a dependency on another file. An
+   ``import module`` declaration cannot induce a dependency.
+
+   If the declaration adds no new bindings into the environment (because, for example,
+   the module name is not used to qualify any identifiers), it is a warning
+   ``-Wredundant-local-import``.
+
+   This declaration form is allowed only with ``-XLocalModules``.
+
+Optional Change Specifications
+------------------------------
+
+Each numbered item in this section can be considered separately.
+
+1. A local module declaration can omit the module name,
+   making an anonymous local module. The names exported
+   by an anonymous module are not added to the enclosing environment qualified,
+   as there is no name to qualify by. (More formally: merge the export
+   environment of the local module with the ambient environment.)
```

Comment:
What is the effect of that on the original name?
```
module Top () where
x = True
module where x = True
```
What are the original names of these two `x`s? Would would `Top.x` refer to?